File: src/service/handle_msg.ts
import type { IUserRepo, User } from "#root/ports/user.js";
import { UserStatus } from "db/enums.js";
import type TelegramBot from "node-telegram-bot-api";
import { randomBytes } from "node:crypto";
import { type AdminService, InvationCmd } from "./admin.js";

interface Config {
	clientConfigEndpoint: string;
}

export class HandleMsgService {
	constructor(
		private readonly bot: TelegramBot,
		private readonly userRepo: IUserRepo,
		private readonly adminService: AdminService,
		private readonly conf: Config,
	) {}

	async handleCallback(msg: TelegramBot.CallbackQuery) {
		if (msg.from?.is_bot) return;
		const isAdminCmd = this.adminService.isAdminCallback(msg);
		if (!isAdminCmd) return;

		await this.adminService.handleAdminCallback(isAdminCmd);
	}

	async handleMsg(msg: TelegramBot.Message) {
		if (msg.from?.is_bot) return;
		const userId = msg.from?.id;
		if (!userId) throw new Error("User ID is required");

		const user = await this.userRepo.select(userId);
		if (!user) {
			await this.register(msg.from!);
		} else {
			await this.sendStatus(user);
		}
	}

	async register(user: TelegramBot.User) {
		const auth_key = randomBytes(32).toString("hex");
		await this.userRepo.insert({
			id: user.id,
			username: user.username || "",
			status: UserStatus.New,
			auth_key,
			created_at: new Date().toISOString(),
		});
		this.sendMessageToAdmin(user, "Новая заявка");
	}

	async sendStatus(user: User) {
		if (user.status == UserStatus.New) {
			this.bot.sendMessage(
				user.id,
				`Администратор скоро рассмотрит вашу заявку`,
			);
			return;
		}
		if (user.status == UserStatus.Accepted) {
			const configURL = this.getConfigLink(user);
			this.bot.sendMessage(
				user.id,
				`Вам одобрен доступ. Ваш файл конфигурации доступен по ссылке: ${configURL}`,
			);
			return;
		}
		if (user.status == UserStatus.Rejected) {
			this.bot.sendMessage(user.id, `Ваша заявка отклонена`);
			return;
		}
	}

	private getConfigLink(user: DB.User) {
		const u = new URL(user.id.toString(), this.conf.clientConfigEndpoint);
		return u.toString();
	}

	private sendMessageToAdmin(user: TelegramBot.User, message: string) {
		const msg = `Новая заявка от ${user.username ?? user.id}: ${message}`;
		this.bot.sendMessage(this.adminService.adminId, msg, {
			reply_markup: {
				inline_keyboard: [
					[
						{
							text: "Accept",
							callback_data: new InvationCmd(
								user.id,
								UserStatus.Accepted,
							).toString(),
						},
						{
							text: "Reject",
							callback_data: new InvationCmd(
								user.id,
								UserStatus.Rejected,
							).toString(),
						},
					],
				],
				one_time_keyboard: true,
				resize_keyboard: true,
			},
		});
	}
}

File: rust/src/service/handle_msg.rs
use crate::db::{enums::UserStatus, NewUser, User};
use crate::ports::user::IUserRepo;
use crate::service::admin::{AdminService, InvitationCmd};
use rand::Rng;
use std::sync::Arc;
use teloxide::prelude::*;
use teloxide::types::{ChatId, InlineKeyboardButton, InlineKeyboardMarkup, Message, User as TgUser};
use teloxide::Bot;
use tracing::info;

pub struct HandleMsgService {
    user_repo: Arc<dyn IUserRepo>,
    admin_service: AdminService,
    client_config_endpoint: String,
}

impl HandleMsgService {
    pub fn new(
        user_repo: Arc<dyn IUserRepo>,
        admin_service: AdminService,
        client_config_endpoint: String,
    ) -> Self {
        Self {
            user_repo,
            admin_service,
            client_config_endpoint,
        }
    }

    pub async fn handle_callback(&self, query: &CallbackQuery) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        if query.from.is_bot {
            return Ok(());
        }

        if let Some(cmd) = self.admin_service.is_admin_callback(query) {
            self.admin_service.handle_admin_callback(&cmd).await?;
        }

        Ok(())
    }

    pub async fn handle_msg(&self, msg: &Message) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let from = msg.from.as_ref();
        if from.is_none() || from.unwrap().is_bot {
            return Ok(());
        }

        let from = from.unwrap();
        let user_id = from.id.0 as i64;
        info!(user_id = user_id, "Handling message");

        let existing_user = self.user_repo.select(user_id)?;

        match existing_user {
            Some(user) => {
                self.send_status(user_id, &user).await?;
            }
            None => {
                self.register(from).await?;
            }
        }

        Ok(())
    }

    async fn register(&self, user: &TgUser) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let auth_key = self.generate_auth_key();
        let new_user = NewUser {
            id: user.id.0 as i64,
            username: user.username.clone().unwrap_or_default(),
            status: UserStatus::New as i32,
            auth_key,
            created_at: chrono::Utc::now().to_rfc3339(),
        };

        self.user_repo.insert(&new_user)?;
        self.send_message_to_admin(user, "Новая заявка").await?;

        Ok(())
    }

    async fn send_status(&self, user_id: i64, user: &User) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let bot = Bot::from_env();
        let chat_id = ChatId(user_id);

        let status = user.status_enum();

        match status {
            UserStatus::New => {
                bot.send_message(chat_id, "Администратор скоро рассмотрит вашу заявку")
                    .await?;
            }
            UserStatus::Accepted => {
                let config_url = self.get_config_link(user);
                bot.send_message(
                    chat_id,
                    format!(
                        "Вам одобрен доступ. Ваш файл конфигурации доступен по ссылке: {}",
                        config_url
                    ),
                )
                .await?;
            }
            UserStatus::Rejected => {
                bot.send_message(chat_id, "Ваша заявка отклонена")
                    .await?;
            }
        }

        Ok(())
    }

    fn get_config_link(&self, user: &User) -> String {
        let base = self.client_config_endpoint.trim_end_matches('/');
        format!("{}/{}", base, user.id)
    }

    fn generate_auth_key(&self) -> String {
        let mut rng = rand::rng();
        let mut bytes = [0u8; 32];
        rng.fill_bytes(&mut bytes);
        hex::encode(bytes)
    }

    async fn send_message_to_admin(
        &self,
        user: &TgUser,
        message: &str,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let bot = Bot::from_env();
        let msg = format!(
            "Новая заявка от {}: {}",
            user.username
                .clone()
                .unwrap_or_else(|| user.id.0.to_string()),
            message
        );

        let accept_cmd = InvitationCmd::new(user.id.0 as i64, UserStatus::Accepted);
        let reject_cmd = InvitationCmd::new(user.id.0 as i64, UserStatus::Rejected);

        let keyboard = InlineKeyboardMarkup::new(vec![vec![
            InlineKeyboardButton::callback("Accept", accept_cmd.to_callback_data()?),
            InlineKeyboardButton::callback("Reject", reject_cmd.to_callback_data()?),
        ]]);

        bot.send_message(ChatId(self.admin_service.admin_id), msg)
            .reply_markup(keyboard)
            .await?;

        Ok(())
    }
}
